odilia/src/tower.rs

```diff
@@ -0,0 +1,258 @@
+#![allow(dead_code)]
+
+use atspi::events::{document::DocumentEvents, object::ObjectEvents};
+use atspi::AtspiError;
+use atspi::Event;
+use atspi::GenericEvent;
+use odilia_common::errors::OdiliaError;
+use std::future::Future;
+use std::marker::PhantomData;
+
+use futures::future::{err, join_all, ok, Either, FutureExt, JoinAll, Ready};
+use futures::{Stream, StreamExt};
+use std::collections::HashMap;
+use std::task::Context;
+use std::task::Poll;
+
+use tower::util::BoxService;
+use tower::Layer;
+use tower::Service;
+
+#[derive(Debug)]
+pub enum Command {
+    Speak(String),
+}
+
+type Response = Vec<Command>;
+type Request = Event;
+type Error = OdiliaError;
+
+pub struct Handlers<S> {
+	state: S,
+	atspi_handlers: HashMap<(String, String), Vec<BoxService<Event, Response, Error>>>,
+  command_handlers: HashMap<Command, Vec<BoxService<Command, (), Error>>>,
+}
+impl<S> Handlers<S>
+where
+	S: Clone + Send + Sync + 'static,
+{
+	pub fn new(state: S) -> Self {
+		Handlers { state, atspi_handlers: HashMap::new(), command_handlers: HashMap::new() }
+	}
+	pub async fn atspi_handler<R>(mut self, mut events: R)
+	where
+		R: Stream<Item = Result<Event, AtspiError>> + Unpin,
+	{
+		std::pin::pin!(&mut events);
+		while let Some(Ok(ev)) = events.next().await {
+			let r = match ev {
+				Event::Object(ObjectEvents::StateChanged(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Object(ObjectEvents::TextCaretMoved(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Object(ObjectEvents::ChildrenChanged(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Object(ObjectEvents::TextChanged(e)) => {
+					self.call_event_listeners(e).await
+				}
+				Event::Document(DocumentEvents::LoadComplete(e)) => {
+					self.call_event_listeners(e).await
+				}
+				_ => {
+					println!("Not implemented yet....");
+					vec![Ok(vec![])]
+				}
+			};
+			for res in r {
+				if let Ok(resp) = res {
+				} else if let Err(err) = res {
+					println!("ERR: {:?}", err);
+				}
+			}
+		}
+	}
+	async fn call_event_listeners<'a, E>(&mut self, ev: E) -> Vec<Result<Response, Error>>
+	where
+		E: atspi::GenericEvent<'a> + Into<Event> + Send + Sync + 'a,
+	{
+		let dn = (
+			<E as atspi::GenericEvent<'a>>::DBUS_MEMBER.into(),
+			<E as atspi::GenericEvent<'a>>::DBUS_INTERFACE.into(),
+		);
+		let input = ev.into();
+    // NOTE: Why not use join_all(...) ?
+    // Because this drives the futures concurrently, and we want ordered handlers.
+    // Otherwise, we cannot guarentee that the caching functions get run first.
+    // we could move caching to a separate, ordered system, then parallelize the other functions,
+    // if we determine this is a performance problem.
+    let mut results = vec![];
+		for hand in self.atspi_handlers.entry(dn).or_default() {
```

Comment:
The advantahe of async here is not that we patallelize on add concurrabce for events, it's that all I/O like which keybinding has been pressesd, writing to SSIP sockets, receicing commands from plugins, etc. can all be done concurrently.

However, the processing of the events and commands must be done in order. There is no way around it if we want to keep cache coherency.

Reply:
cash shouldn't be treated as a source of truth, the tree should. As such, if the race condition happens and an item is referenced before it's added to cache, well, do so then. Cache should also be invalidated from time to time, because the point is not to work off of the cache in most cases anyway, or that's how I understand the way caches work at least. Look it up in cache first, but if it's not there, we shouldn't have issues with just querying the bus, even if it'll be added literally with the next handler call, the next time this is called will have it in the cache then. Or am I misunderstanding what you're refering to?

Reply:
The cache is the source of truth. Whenever an update comes in over the bus, we update the cache; if an item is not found in the cache, we fetch it and all its data.

For an example, if we request something like this in an eventual handler:

```rust
async fn read_focused_item(
    FocusedItem(item): FocusedItem,
) -> Vec<Command> {
    vec![Speak(format!("{}, {}", item.text, item.role).into()]
}
```

Assuming we handle all updates correctly, we should be able to keep a fairly accurate accounting of the a11y tree within the cache.

Reply:
well, then this isn't only a cache, perhaps it should be renamed down the line to clear confusion. Maybe a11y tree database or something like that? cache implies something being temporary and only checked there for speed, and not a source of truth on anything, especially since most caches invalidate themselves eventually

Reply:
Right, but most caches are also transparent as to whether they have the cached item or not. So if an item is not found, the cache usually loads it into the cache immediately, without intervention from the user. AFAIK anyway.